src/wlserver.cpp

```diff
@@ -1083,6 +1084,208 @@ static void create_gamescope_control( void )
 	wl_global_create( wlserver.display, &gamescope_control_interface, version, NULL, gamescope_control_bind );
 }
 
+////////////////////////
+// commit-timing
+////////////////////////
+
+enum timer_stage {
+	TIMER_STAGE_LATCH = 0,
+	TIMER_STAGE_PRESENT = 1
+};
+
+enum timer_rounding_mode {
+	TIMER_ROUNDING_NEAREST = 0,
+	TIMER_ROUNDING_NOT_BEFORE = 1
+};
+
+struct commit_timer {
+	struct wl_resource *resource;
+
+	struct wlr_surface *surface;
+	enum timer_stage curr_stage;
+	enum timer_rounding_mode curr_rounding_mode;
+	uint64_t curr_timestamp_nsec;
+
+	struct wl_listener client_commit;
+	struct wlr_addon surface_addon;
+};
+
+static inline int64_t timespec_to_nsec(const struct timespec *a) {
+	return (int64_t)a->tv_sec * 1'000'000'000ul + a->tv_nsec;
+}
+
+static void resource_handle_destroy(struct wl_client *client, struct wl_resource *resource) {
+       wl_resource_destroy(resource);
+}
+
+static void commit_timer_destroy(struct commit_timer *timer) {
+	if (timer == NULL) {
+		return;
+	}
+	wl_list_remove(&timer->client_commit.link);
+	wlr_addon_finish(&timer->surface_addon);
+	wl_resource_set_user_data(timer->resource, NULL); // make inert
+	free(timer);
+}
+
+static void surface_timer_addon_handle_destroy(struct wlr_addon *addon) {
+	struct commit_timer *timer = wl_container_of(addon, timer, surface_addon);
+	commit_timer_destroy(timer);
+}
+
+static const struct wlr_addon_interface surface_timer_addon_impl = {
+	.name = "wp_commit_timer_v1",
+	.destroy = surface_timer_addon_handle_destroy,
+};
+
+#define NANOS_IN_SEC 1000000000
+static uint64_t timer_get_target_present_nsec(uint64_t timestamp_nsec, uint32_t rounding_mode) {
+	uint64_t refresh_nsec = NANOS_IN_SEC/(GetVBlankTimer().GetRefresh()/1000);
+	uint64_t cycle_phase_nsec = GetVBlankTimer().GetLastVBlank() % refresh_nsec;
+
+	uint64_t target_nsec = timestamp_nsec;
+	target_nsec -= cycle_phase_nsec;
+	if (rounding_mode == TIMER_ROUNDING_NEAREST) {
+		target_nsec += refresh_nsec/2;
+		target_nsec -= (target_nsec % refresh_nsec);
+	}
+	else if (rounding_mode == TIMER_ROUNDING_NOT_BEFORE) {
+		target_nsec = ((target_nsec+(refresh_nsec-1))/refresh_nsec)*refresh_nsec;
+	}
+	target_nsec += cycle_phase_nsec;
+
+	return target_nsec;
+}
+
+static bool is_valid_timestamp(const struct commit_timer *timer, uint64_t target_present_nsec) {
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	uint64_t now_nsec = timespec_to_nsec(&now);
+	if (target_present_nsec <= now_nsec)
+		return false;
+
+	std::vector<ResListEntry_t>* commits;
+	if (auto& x11_info = get_wl_surface_info(timer->surface)->x11_surface)
+		commits = &x11_info->xwayland_server->retrieve_commits();
+	else if (get_wl_surface_info(timer->surface)->xdg_surface)
+		commits = &wlserver.xdg_commit_queue;
+	else
+		return true;
+
+	for (const auto& commit : *commits)
+	{
+		if (commit.desired_present_time > target_present_nsec)
```

Comment:
This logic doesn't work because some pending commits have already probably moved to steamcompmgr already.

Unfortunately Gamescope is not set up to really allow you to check something like this given it's architecture.
I am happy to just drop the implementation of the protocol error for now.

Reply:
Actually, you could just store the highest used target_present_nsec on the wlserver_wl_surface_info instead of checking every commit.

Reply:
You should also probably do that for your actual wlroots implementation instead of checking over every commit ;)

Reply:
Huh, nice...
From now on whenever I find myself mindlessly proliferating cpu cycles, I'll ask myself if I could do any better :D

However I think I'll just leave a very trivial check for now, since [it appears to be controversial what to actually consider an error in this case](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/248#note_2378990).